src/storage/store/engine.rs

```diff
@@ -203,30 +196,42 @@ impl ShardEngine {
         &mut self,
         max_wait: Duration,
     ) -> Result<Vec<MempoolMessage>, EngineError> {
-        let mut messages = Vec::new();
-        let start_time = Instant::now();
+        if let Some(messages_request_tx) = &self.messages_request_tx {
+            let mut messages = Vec::new();
+            let start_time = Instant::now();
 
-        loop {
-            if start_time.elapsed() >= max_wait {
-                break;
-            }
+            loop {
+                if start_time.elapsed() >= max_wait {
+                    break;
+                }
+
+                while messages.len() < self.max_messages_per_block as usize {
+                    let (message_tx, message_rx) = oneshot::channel();
+
+                    if let Err(err) = messages_request_tx.send((self.shard_id, message_tx)).await {
+                        error!(
+                            "Could not send request for messages to mempool {}",
+                            err.to_string()
+                        )
+                    }
 
-            while messages.len() < self.max_messages_per_block as usize {
-                match self.messages_rx.try_recv() {
-                    Ok(msg) => messages.push(msg),
-                    Err(mpsc::error::TryRecvError::Empty) => break,
-                    Err(err) => return Err(EngineError::from(err)),
+                    match message_rx.await {
```

Comment:
I think we should sending in the max_messages_per_block and get all the messages in one go. You can also combine it with the timeout function to get rid of the loop https://docs.rs/tokio/latest/tokio/time/fn.timeout.html

Reply:
Agreed-- changed it so we return a bulk set of messages in a single response. 

I just left the loop in for now. I think in order for the timeout to work, we need to pass a timeout into the mempool via the messages request and have the mempool timeout the request after the timeout has been hit. Is that what you were imagining? I don't think we'll get the behavior we want with a timeout on the engine side.